job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala

```diff
@@ -0,0 +1,217 @@
+
+      var contextName = ""
+      do {
+        contextName = java.util.UUID.randomUUID().toString().substring(0, 8) + "-" + classPath
+      } while (contexts contains contextName)
+
+      startContext(contextName, mergedConfig, true) { ref =>
+        originator ! (contexts(contextName), resultActors(contextName))
+      } { err =>
+        originator ! ContextInitError(err)
+      }
+
+    case GetResultActor(name) =>
+      sender ! resultActors.get(name).getOrElse(globalResultActor)
+
+    case GetContext(name) =>
+      if (contexts contains name) {
+        sender ! (contexts(name), resultActors(name))
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case StopContext(name) =>
+      if (contexts contains name) {
+        logger.info("Shutting down context {}", name)
+
+        context.watch(contexts(name))
+        contexts(name) ! PoisonPill
+        resultActors.remove(name)
+        sender ! ContextStopped
+      } else {
+        sender ! NoSuchContext
+      }
+
+    case Terminated(actorRef) =>
+      val name: String = actorRef.path.name
+      logger.info("Actor terminated: {}", name)
+      contexts.foreach { kv => if (kv._2 == actorRef) contexts.remove(kv._1) }
+      resultActors.foreach { kv => if (kv._2 == actorRef) resultActors.remove(kv._1) }
+  }
+
+  private def initContext(actorName: String, ref: ActorRef, timeoutSecs: Long = 1)
+                         (ctxName: String, ctxConf: Config, isAdHoc: Boolean)
+                         (successFunc: ActorRef => Unit, failureFunc: Throwable => Unit): Unit = {
+    import akka.pattern.ask
+
+    val resultActor = if (isAdHoc) globalResultActor else context.actorOf(Props(classOf[JobResultActor]))
+    (ref ? JobManagerActor.Initialize(
+      daoActor, Some(resultActor), ctxName, ctxConf, isAdHoc, self))(Timeout(timeoutSecs.second)).onComplete {
+      case Failure(e:Exception) =>
+        logger.info("Failed to send initialize message to context " + ref, e)
+        ref ! PoisonPill
+        failureFunc(e)
+      case Success(JobManagerActor.InitError(t)) =>
+        logger.info("Failed to initialize context " + ref, t)
+        ref ! PoisonPill
+        failureFunc(t)
+      case Success(JobManagerActor.Initialized(_)) =>
+        logger.info("SparkContext {} joined", ctxName)
+        contexts(ctxName) = ref
+        resultActors(ctxName) = resultActor
+        successFunc(ref)
+    }
+
+  }
+
+  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
+                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
+    require(!(contexts contains name), "There is already a context named " + name)
+    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
+
+    logger.info("Starting context with actor name {} ", contextActorName)
+
+    val pb = Process(s"$managerStartCommand $contextActorName ${selfAddress.toString}")
```

Comment:
Hmm I like the actor subtree idea.  I’ll see what I can do, after responding to other generic feedback first and cleaning up logging.

> On Dec 20, 2015, at 10:41 AM, Ankit Soni <notifications@github.com> wrote:
> 
> In job-server/src/spark.jobserver/AkkaClusterSupervisorActor.scala <https://github.com/spark-jobserver/spark-jobserver/pull/345#discussion_r48106654>:
> 
> > +        logger.info("SparkContext {} joined", ctxName)
> > +        contexts(ctxName) = ref
> > +        resultActors(ctxName) = resultActor
> > +        successFunc(ref)
> > +    }
> > +
> > +  }
> > +
> > +  private def startContext(name: String, contextConfig: Config, isAdHoc: Boolean)
> > +                          (successFunc: ActorRef => Unit)(failureFunc: Throwable => Unit) = {
> > +    require(!(contexts contains name), "There is already a context named " + name)
> > +    val contextActorName = "jobManager-" + java.util.UUID.randomUUID().toString.substring(16)
> > +
> > +    logger.info("Starting context with actor name {} ", contextActorName)
> > +
> > +    val pb = Process(s"$managerStartCommand $contextActorName ${selfAddress.toString}")
> Oh ok, that makes sense! I wasn't sure what exactly HA means here. If there is a global view of contexts, it should not be possible to have multiple contexts with the same name (atleast after reconciling different masters views) if i understand correctly .
> 
> Anyways, just wanted to add that even without multiple masters, the problem of identifying the JobManager actors still stands because there are other actors in the system. Maybe a more idiomatic way to do it is have the JobManagerActors be children of a "contexts" actor - so their paths would be <actorSystem>/user/contexts/context1, etc. and the akka supervisor will then send identify messages only to that subtree <actorSystem>/user/contexts/* when a new member joins.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/spark-jobserver/spark-jobserver/pull/345/files#r48106654>.
>